Adding an optional array of server vars, to make it possible for auto authentication to occur during basic HTTP auth.

brandonsavage opened the pull request:
This fixes an issue, which is that the user currently has to deal with the $_SERVER variables when doing basic auth. Since basic auth can happen all over the application and the user is rarely redirected to an auth page, this will simplify the process of authenticating a user.

If an application is using http auth, and they receive an anonymous user object, they can automatically attempt to log the user in with the credentials that were supplied by the user. These credentials can be automatically loaded into the verification mechanism, avoiding having to do this in each and every action or controller.

brandonsavage committed:
Adding an optional array of server vars, to make it possible for auto authentication to occur during basic HTTP auth.

brandonsavage committed:
Making patch PHP 5.3 compliant.

harikt commented on src/Adapter/HtpasswdAdapter.php:
```diff
@@ -88,11 +117,21 @@ public function getVerifier()
-        $username = $input['username'];
-        $password = $input['password'];
+
+        $username = $this->username;
```
@brandonsavage I believe this change will break the default expectation of username and password given.

brandonsavage replied:
These will both be overridden if the user provides them as part of the input. They won't break anything.

harikt replied:
aah just noticed it.

Nice :) .

pmjones commented:
Been thinking about this. You're right that "basic auth can happen all over the application", and having the $_SERVER info as a constructor param is useful in that case. However, we can remove the "basic" from your sentence, and it's still true: "auth [of any kind] can happen all over the application".

Maybe I'm generalizing the case too broadly, but I wouldn't want to have this functionality for every adapter in the package.  I don't think it's _too_ onerous for the calling code to receive the $_SERVER values, and pass them along to the adapter in the way the adapter expects. That keeps it consistent for every adapter.

Thoughts?

brandonsavage commented:
I think that's reasonable.

It's worth noting that in my initial interpretation I confused how this was
implemented. I initially thought that you were using Apache's basic auth
system. That turns out not to be the case. Instead, you're using a
username/password field and authenticating the user against the password
file. That works just as well and I've implemented that in several small
apps I have.

On Mon, Mar 16, 2015 at 6:23 PM, Paul M. Jones notifications@github.com
wrote:

> Been thinking about this. You're right that "basic auth can happen all
> over the application", and having the $_SERVER info as a constructor param
> is useful in that case. However, we can remove the "basic" from your
> sentence, and it's still true: "auth [of any kind] can happen all over the
> application".
> 
> Maybe I'm generalizing the case too broadly, but I wouldn't want to have
> this functionality for every adapter in the package. I don't think it's
> _too_ onerous for the calling code to receive the $_SERVER values, and
> pass them along to the adapter in the way the adapter expects. That keeps
> it consistent for every adapter.
> 
> Thoughts?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/auraphp/Aura.Auth/pull/57#issuecomment-81965992.

pmjones commented:
Excellent. Thanks for the followup. :-)
